awx/main/scheduler/dag_workflow.py

```diff
@@ -63,40 +112,113 @@ def cancel_node_jobs(self):
+                failed_nodes.append(node)
+
+        for node in failed_nodes:
+            obj = node['node_object']
+            if (len(self.get_dependencies(obj, 'failure_nodes')) +
+                    len(self.get_dependencies(obj, 'always_nodes'))) == 0:
+                if obj.unified_job_template is None:
+                    res = True
+                    failed_unified_job_template_node_ids.append(str(obj.id))
                 else:
-                    is_failed = True if children_all else job.status in ['failed', 'canceled', 'error']
+                    res = True
+                    failed_path_nodes_id_status.append((str(obj.id), obj.job.status))
 
-            if job.status in ['canceled', 'error']:
-                continue
-            elif job.status == 'failed':
-                nodes.extend(children_failed + children_always)
-            elif job.status == 'successful':
-                nodes.extend(children_success + children_always)
+        if res is True:
+            s = _("No error handle path for workflow job node(s) [{node_status}] workflow job "
+                  "node(s) missing unified job template and error handle path [{no_ufjt}].")
+            parms = {
+                'node_status': '',
+                'no_ufjt': '',
+            }
+            if len(failed_path_nodes_id_status) > 0:
+                parms['node_status'] = ",".join(["({},{})".format(id, status) for id, status in failed_path_nodes_id_status])
+            if len(failed_unified_job_template_node_ids) > 0:
+                parms['no_ufjt'] = ",".join(failed_unified_job_template_node_ids)
+            return True, smart_text(s.format(**parms))
+        return False, None
+
+    r'''
+    Determine if all nodes have been decided on being marked do_not_run.
+    Nodes that are do_not_run False may become do_not_run True in the future.
+    We know a do_not_run False node will NOT be marked do_not_run True if there
+    is a job run for that node.
+
+    :param workflow_nodes:     list of workflow_nodes
+
+    Return a boolean
+    '''
+    def _are_all_nodes_dnr_decided(self, workflow_nodes):
+        for n in workflow_nodes:
+            if n.do_not_run is False and not n.job and n.unified_job_template:
+                return False
+        return True
+
+
+    r'''
+    Determine if a node (1) is ready to be marked do_not_run and (2) should
+    be marked do_not_run.
+
+    :param node:             SimpleDAG internal node
+    :param parent_nodes:     list of workflow_nodes
+
+    Return a boolean
+    '''
+    def _should_mark_node_dnr(self, node, parent_nodes):
+        for p in parent_nodes:
+            if p.do_not_run is True:
+                pass
+            elif p.job:
+                if p.job.status == 'successful':
+                    if node in (self.get_dependencies(p, 'success_nodes') +
+                                self.get_dependencies(p, 'always_nodes')):
```

Comment:
This logic may not be reliable if multiple connection types are set. `p` may be _both_ a success node _and_ an always node.

I would think this method would need to be sensitive to the connection type for a given parent node, allowing that a node may appear more than once in the `parent_nodes` list.

Will file new issue. It's probably better to avoid to prohibiting that action.

Reply:
https://github.com/ansible/awx/issues/2782